leetcodePython/matrix/matrix_2812.py

```diff
@@ -0,0 +1,73 @@
+import collections
+from typing import List
+from heapq import *
+
+
+class Solution:
+    def maximumSafenessFactor(self, grid: List[List[int]]) -> int:
+        directions = [(0, 1), (0, -1), (1, 0), (-1, 0)]
+
+        grid_size = len(grid)
+        distance = [[float('inf')] * grid_size for _ in range(grid_size)]
+        pos_queue = collections.deque()
+        self._record_thief_pos(grid, pos_queue, distance)
+        self._set_thief_distance(pos_queue, distance, directions)
+        safeness_max_heap = [(-distance[0][0], 0, 0)]
+        max_safeness: List[List[float]]
+        max_safeness = [[-1] * grid_size for _ in range(grid_size)]
+        max_safeness[0][0] = distance[0][0]
+
+        while safeness_max_heap:
+            neg_safe, r, c = heappop(safeness_max_heap)
+            safeness = - neg_safe
+            if r == grid_size - 1 and c == grid_size - 1:
+                return int(safeness)
+            for dr, dc in directions:
+                nr, nc = r + dr, c + dc
+                if self._out_of_bound(grid_size, nr, nc):
+                    continue
+                min_safeness_dir = min(safeness, distance[nr][nc])
```

Comment:
When we hit this point. It's path safeness won't be bigger than:
* It's distance to thief 
* Safeness of previous point.